### ISSUE TITLE
OpenGL examples: save/restore viewport state.

### ISSUE THREAD

**Issue by javawag** [ocornut/imgui#392]
Hi,

I've integrated imGui into my game engine (which uses SDL2 & OpenGL 3/no fixed pipeline), but I experienced some horrible graphical glitches.

Turns out this is because I'm using glViewport() to letterbox the viewport, to maintain the correct aspect ratio. The fix to get around this is, similar to the existing GL state caching, is to use this:

```C++
ImGuiIO const &io = ImGui::GetIO();

// Store VP
GLint aiViewport[4];
glGetIntegerv(GL_VIEWPORT, aiViewport);
glViewport(0, 0, (GLsizei)io.DisplaySize.x, (GLsizei)io.DisplaySize.y);

/* ...render ImGui...  */

// Restore VP
glViewport(aiViewport[0], aiViewport[1], (GLsizei)aiViewport[2], (GLsizei)aiViewport[3]);
```

I think this is syntax-correct but I don't have access to my IDE right now.

Cheers!

**Comment by ocornut** [comment]
I'm not sure it should be the role of this code is set the viewport based on io.DisplaySize. Wouldn't it make more sense that you code set the viewport itself prior to calling Render?
(though to be fair if the user wants to do something unusual they will probably modify imgui_impl_* files to alter the projection matrix, anyway in which case adding glViewport may not be an issue because it can be removed)

By "horrible graphic glitches" you mean that ImGui was squashed or clamped?

**Comment by javawag** [comment]
Perhaps. Certainly, that's how I've implemented it myself. I just wondered, since in the example code most other GL state is cached and later restored, if this should be another case which is handled in the example code.

And yes, to clarify - ImGui looked fine if I sized the window so that there was no letterboxing. However, the further the aspect ratio of the scene got from the aspect ratio of the window (i.e. the larger the black bars at the top/bottom or sides), the more squashing/clamping/weird spacing occurred.
